src/linear_algebra/multilinear.lean

```diff
@@ -515,6 +516,135 @@ protected def pi_ring_equiv [fintype ι]  : M₂ ≃ₗ[R] (multilinear_map R (
 
 end comm_ring
 
+section algebra
+
+variables (S : Type u) [comm_semiring S]
+variables {N : Type u} [semiring N] [algebra S N]
+variable {q : ℕ}
+variable (ν : fin q → N)
+
+@[priority 100] instance algebra_to_semimodule : semimodule S N := algebra.to_semimodule
+
+/--
+Auxiliary function used to define the multilinear map `algebra_prod`.
+-/
+def algebra_prod_aux : N := list.prod $ list.map ν (list.fin_range q)
```

Comment:
It is often useful to use a general `fintype` rather than `fin` as an index for finite-dimensional vector spaces, so that would hypothetically look something like:
```suggestion
variable {q : Type*} [fintype q]
variable (ν : q → N)

@[priority 100] instance algebra_to_semimodule : semimodule S N := algebra.to_semimodule

/--
Auxiliary function used to define the multilinear map `algebra_prod`.
-/
def algebra_prod_aux : N := finset.univ.prod q
```
However, this would require that `N` is a `comm_semiring`, not just a `semiring`. Are there non-commutative examples of `N` that might be used? (As far as I know, exterior products are generally defined over a field, so my impression is that we can assume it is commutative.)

Reply:
If the order of multiplication matters, maybe take the list of indices as a parameter:
```suggestion
variable {q : Type*} (qs : list q)
variable (ν : q → N)

@[priority 100] instance algebra_to_semimodule : semimodule S N := algebra.to_semimodule

/--
Auxiliary function used to define the multilinear map `algebra_prod`.
-/
def algebra_prod_aux : N := list.prod $ list.map ν qs
```
Apart from making the types more general, I expect that this can simplify the proofs below, because we can do the `fin_range` and the `prod` and `map` parts separately.

Reply:
What if I worked with a linearly ordered fintype?

Reply:
The non-commutative examples of N are the important ones, since the same file already deals with the commutative case.

Reply:
If I do your second suggestion, can my proofs by induction on `q` be converted easily?

Reply:
Then indeed I would use the second suggestion. You should be able to do the start off by induction on the list (instead of the size of the list that it was (arguably) before). For example, the start of `algebra_prod` might look like:

```lean
def algebra_prod : multilinear_map S (λ i : q, N) N :=
{ to_fun := algebra_prod_aux,
  map_add' :=
  begin
    intros ν i x y,
    induction qs with head tail hq,
```

Reply:
Sorry for not responding, I was busy this weekend and didn't check my notifications enough.

1. If you think that works better, of course use it! I originally suggested a list of indices to make the analogy to the commutative `finset.sum` easier. I do not know if the rest might be easier for either way. (In my experience, I can only know that I made the right choices in a definition after I use it in proofs, before that I need to rely on intuition.)

2. I appreciate that the type `list N` is more "native" to functional programming than linear algebra, compared to `fin q -> N`. For the mathlib project, this is not an important consideration though. Contributions should use less beginner-friendly concepts if it makes the theory easier to work with once you learn them, and anyone is very much invited to ask questions to the community. (The (in)famous example is using filters to define limits of sequences and functions at once.) A different interface for the definitions can be useful, but I would say that would be part of future work, after this pull request.

I would like to make clear that I would still like to see the definition of `algebra_prod` and am already grateful for the original definition in terms of `fin q -> N`. Let's get it perfect before merging it, though.

Reply:
With the push I just did, these `aux` definitions are completely gone, as they duplicate `multilinear_map.mk_pi_algebra_fin`